src/common/device_helpers.cuh

```diff
@@ -752,6 +752,29 @@ void Gather(int device_idx, T *out, const T *in, const int *instId, int nVals) {
                                        });
 }
 
+class SaveCudaContext {
```

Comment:
What about calling it `WithDevice` and passing the new device number in? In this way, it can save the old device and switch to the new one, and then switch back in the destructor. Ideally, this would enable avoiding calling `cudaSetDevice` directly in xgboost code.

Also, what about making it a function instead of a class?

Reply:
I will leave this to @trivialfis

Reply:
Ideally,  we should avoid changing device in master thread.  So maybe working toward that end would be better.